src/mapgen/mg_schematic.cpp

```diff
@@ -436,9 +436,20 @@ bool Schematic::serializeToLua(std::ostream *os,
 				u8 probability   = schemdata[i].param1 & MTSCHEM_PROB_MASK;
 				bool force_place = schemdata[i].param1 & MTSCHEM_FORCE_PLACE;
 
-				ss << indent << indent << "{"
-					<< "name=\"" << names[schemdata[i].getContent()]
-					<< "\", prob=" << (u16)probability * 2
+				content_t c = schemdata[i].getContent();
+
+				ss << indent << indent << "{" << "name=\"";
+
+				if (c < names.size()) {
+					// Get node name from cache
+					ss << names[c];
+				} else {
+					// If cache is unavailable (biome decoration)
+					FATAL_ERROR_IF(m_ndef == nullptr, "Invalid node list and missing NodeDefManager");
+					ss << m_ndef->get(c).name;
```

Comment:
IMO it'd be okay to fill up names again once at the beginning of `serializeToLua` if it's detected that they are missing

Reply:
I could copy the `names` vector and rebuild `m_nodenames` but these serialization functions are currently designed to take an external list of node names rather than relying on their own.

Reply:
Oh I thought the names vector belonged to the `Schematic` class. This is weird.
Who calls `serializeToLua`? Can we do it there?

Reply:
When this case is reached, the only available content_t -> node name conversion is NodeDefResolver. The other case is a "condensed" (sequential numbered) map of node IDs which requires m_nodenames to look up the names.